### PR TITLE
Gas Optimizations (Pump)

### PR DESCRIPTION
**Pull request by pynchmeister**
# Gas Optimizations (Pump)

Made several gas optimizations including:

* [Left shift instead of multiplying by powers of two saves gas](https://github.com/pynchmeister/Wells/commit/49a159a3e46b50009c7d750168da0a7f0edc2a8c)

* Cache array length during for loop definition.

### CODE REVIEW

File: src/pumps/GeoEmaAndCumSmaPump.sol [hunk 1]
```diff
@@ -282,7 +286,7 @@ contract GeoEmaAndCumSmaPump is IPump, IInstantaneousPump, ICumulativePump {
      * @dev Get the starting byte of the slot that contains the `n`th element of an array.
      */
     function getSlotsOffset(uint n) internal pure returns (uint) {
-        return ((n - 1) / 2 + 1) * 32; // Maybe change to n * 32?
+        return ((n - 1) / 2 + 1) << 32; // Maybe change to n * 32?
```
Comment (Brean0) [inline src/pumps/GeoEmaAndCumSmaPump.sol:289]: should this be << 5?

File: src/pumps/GeoEmaAndCumSmaPump.sol [hunk 2]
```diff
@@ -126,7 +126,8 @@ contract GeoEmaAndCumSmaPump is IPump, IInstantaneousPump, ICumulativePump {
         bytes16[] memory byteReserves = new bytes16[](reserves.length);
 
         // Skip {_capReserve} since we have no prior reference
-        for (uint i = 0; i < reserves.length; i++) {
+        uint length = reserves.length;
```
Comment (Brean0) [inline src/pumps/GeoEmaAndCumSmaPump.sol:129]: can you put this assignment above line 126 so that bytesReserve can also use length?

**Review by Brean0 — CHANGES_REQUESTED** [review]

File: src/pumps/GeoEmaAndCumSmaPump.sol (hunk 2 above)
Comment (pynchmeister) [reply src/pumps/GeoEmaAndCumSmaPump.sol:129]: yes no doubt

File: src/pumps/GeoEmaAndCumSmaPump.sol (hunk 1 above)
Comment (pynchmeister) [reply src/pumps/GeoEmaAndCumSmaPump.sol:289]: right on, fixed.

**Review by Brean0 — APPROVED** [review]

File: src/pumps/GeoEmaAndCumSmaPump.sol [hunk 3]
```diff
@@ -240,7 +245,7 @@ contract GeoEmaAndCumSmaPump is IPump, IInstantaneousPump, ICumulativePump {
     function _readCumulativeReserves(address well) internal view returns (bytes16[] memory cumulativeReserves) {
         bytes32 slot = getSlotForAddress(well);
         (uint8 n, uint40 lastTimestamp, bytes16[] memory lastReserves) = slot.readLastReserves();
-        uint offset = getSlotsOffset(n) * 2;
+        uint offset = getSlotsOffset(n) << 2;
```
Comment (Brean0) [inline src/pumps/GeoEmaAndCumSmaPump.sol:248]: missed this, this should be shifted by 1

Comment (pynchmeister) [reply src/pumps/GeoEmaAndCumSmaPump.sol:248]: Good catch, fixed

**Pull request merged**